### PR TITLE
Don't panic on non-UTF-8 arguments; just fail

### LINKED ISSUE #28: getopts panics parsing non-UTF-8 arguments
Arguments are `OsString` objects. While `parse()` takes an `IntoIterator<Item=AsRef<OsStr>>` (which can safely accept `env::args_os()`), the `Matches` structure only gives out `String` objects, which can only be built for UTF-8.

I don't know if this can possibly be fixed without breaking anything, but new safe methods on `Matches` could certainly be introduced.

### PR DESCRIPTION
**Pull request by remram44**
["plugging the panics"](https://github.com/rust-lang/getopts/pull/29#issuecomment-119259958) as requested

Fixes #28

### CODE REVIEW

**Comment by remram44** [comment]
I still think getopts should be able to parse non-utf-8 OsString arguments (like Paths).

File: src/lib.rs [hunk 1]
```diff
@@ -304,9 +304,20 @@ impl Options {
...
-        }).collect::<Vec<_>>();
+        let args = {
+            let iter = args.into_iter();
+            let mut vec = Vec::with_capacity(iter.size_hint().0);
+            for i in iter {
+                match i.as_ref().to_str() {
+                    Some(s) => vec.push(s.to_owned()),
+                    None => {
+                        let readable = format!("{:?}", i.as_ref());
+                        return Err(Fail::UnrecognizedOption(readable))
```
Comment (remram44) [inline src/lib.rs]: There could be a specific Fail variant for this I guess? I went the compatible way.

File: src/lib.rs [hunk 2]
```diff
@@ -304,9 +304,20 @@ impl Options {
...
+                    }
+                }
+            }
+            vec
+        };
```
Comment (alexcrichton) [inline src/lib.rs]: Using `with_capacity` isn't really necessary here, so I thinks something like this should be ok:

```rust
let args = try!(args.into_iter().map(|i| {
    i.as_ref().to_str().ok_or_else(|| {
        Fail::UnrecognizedOption(format!("{:?}", i.as_ref()))
    })
}).collect::<Result<Vec<_>, _>());
```

**Comment by alexcrichton** [comment]
> I still think getopts should be able to parse non-utf-8 OsString arguments (like Paths).

I agree! This can't be done without making breaking changes to the API today, however, though.

File: src/lib.rs (hunk 2 above)
Comment (remram44) [inline src/lib.rs]: This is shorter but I'm not sure if it's more readable :smile: It needs a `to_owned()` somewhere, I'll push it in a minute.

**Comment by alexcrichton** [comment]
Thanks!

**Pull request merged**